source/common/config/delta_subscription_impl.cc

```diff
@@ -1,106 +1,74 @@
-  state_->updateResourceInterest(update_to_these_names);
-  // Tell the server about our new interests, if there are any.
-  trySendDiscoveryRequests();
+DeltaSubscriptionImpl::~DeltaSubscriptionImpl() {
+  if (watch_) {
+    context_->removeWatch(type_url_, watch_);
+  }
 }
 
-// Config::GrpcStreamCallbacks
-void DeltaSubscriptionImpl::onStreamEstablished() {
-  state_->markStreamFresh();
-  trySendDiscoveryRequests();
-}
+void DeltaSubscriptionImpl::pause() { context_->pause(type_url_); }
 
-void DeltaSubscriptionImpl::onEstablishmentFailure() { state_->handleEstablishmentFailure(); }
+void DeltaSubscriptionImpl::resume() { context_->resume(type_url_); }
 
-void DeltaSubscriptionImpl::onDiscoveryResponse(
-    std::unique_ptr<envoy::api::v2::DeltaDiscoveryResponse>&& message) {
-  ENVOY_LOG(debug, "Received gRPC message for {} at version {}", type_url_,
-            message->system_version_info());
-  kickOffAck(state_->handleResponse(*message));
+// Config::DeltaSubscription
+void DeltaSubscriptionImpl::start(const std::set<std::string>& resources) {
+  // ADS initial request batching relies on the users of the GrpcMux *not* calling start on it,
+  // whereas non-ADS xDS users must call it themselves.
+  if (!is_aggregated_) {
+    context_->start();
+  }
+  watch_ = context_->addOrUpdateWatch(type_url_, watch_, resources, *this, init_fetch_timeout_);
+  stats_.update_attempt_.inc();
 }
 
-void DeltaSubscriptionImpl::onWriteable() { trySendDiscoveryRequests(); }
-
-void DeltaSubscriptionImpl::kickOffAck(UpdateAck ack) {
-  ack_queue_.push(ack);
-  trySendDiscoveryRequests();
+void DeltaSubscriptionImpl::updateResourceInterest(
+    const std::set<std::string>& update_to_these_names) {
+  watch_ = context_->addOrUpdateWatch(type_url_, watch_, update_to_these_names, *this,
+                                      init_fetch_timeout_);
+  stats_.update_attempt_.inc();
 }
 
-// Checks whether external conditions allow sending a DeltaDiscoveryRequest. (Does not check
-// whether we *want* to send a DeltaDiscoveryRequest).
-bool DeltaSubscriptionImpl::canSendDiscoveryRequest() {
-  if (state_->paused()) {
-    ENVOY_LOG(trace, "API {} paused; discovery request on hold for now.", type_url_);
-    return false;
-  } else if (!grpc_stream_.grpcStreamAvailable()) {
-    ENVOY_LOG(trace, "No stream available to send a DiscoveryRequest for {}.", type_url_);
-    return false;
-  } else if (!grpc_stream_.checkRateLimitAllowsDrain()) {
-    ENVOY_LOG(trace, "{} DiscoveryRequest hit rate limit; will try later.", type_url_);
-    return false;
-  }
-  return true;
+// Config::SubscriptionCallbacks
+void DeltaSubscriptionImpl::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const std::string& version_info) {
+  callbacks_.onConfigUpdate(resources, version_info);
+  stats_.update_success_.inc();
+  stats_.update_attempt_.inc();
+  stats_.version_.set(HashUtil::xxHash64(version_info));
```

Comment:
Why do we set the hash to stats_.version_? It doesn't seems informative, calculating xxHash64 from version_info outside of Envoy isn't trivial.

Reply:
Yeah, it is unfortunate that we have to do this. version_info is a string, but the stats can only report numbers.